### PR TITLE
[SPARK-18726][SQL]resolveRelation for FileFormat DataSource don't need to  listFiles twice

### PR DESCRIPTION
**Pull request by windpiger**
## What changes were proposed in this pull request?

Currently when we resolveRelation for a `FileFormat DataSource` without providing user schema, it will execute `listFiles`  twice in `InMemoryFileIndex` during `resolveRelation`.

This PR add a `FileStatusCache` for DataSource, this can avoid listFiles twice.

But there is a bug in `InMemoryFileIndex` see:
 [SPARK-19748](https://github.com/apache/spark/pull/17079)
 [SPARK-19761](https://github.com/apache/spark/pull/17093), 
so this pr should be after SPARK-19748/ SPARK-19761.

## How was this patch tested?
unit test added

### CODE REVIEW

**Comment by windpiger** [comment]
retest this please

**Comment by windpiger** [comment]
retest this please

**Comment by windpiger** [comment]
cc @cloud-fan @gatorsmile  could you help to review this?thanks :)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 1]
```diff
@@ -364,7 +364,8 @@ case class DataSource(
             catalogTable.get,
             catalogTable.get.stats.map(_.sizeInBytes.toLong).getOrElse(defaultTableSize))
         } else {
-          new InMemoryFileIndex(sparkSession, globbedPaths, options, Some(partitionSchema))
+          new InMemoryFileIndex(sparkSession, globbedPaths, options, Some(partitionSchema),
+            fileStatusCache)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Nit: indent issue

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 2]
```diff
@@ -364,7 +364,12 @@ case class DataSource(
             catalogTable.get,
             catalogTable.get.stats.map(_.sizeInBytes.toLong).getOrElse(defaultTableSize))
         } else {
-          new InMemoryFileIndex(sparkSession, globbedPaths, options, Some(partitionSchema))
+          new InMemoryFileIndex(
+            sparkSession,
+            globbedPaths,
+            options,
+            Some(partitionSchema),
+            fileStatusCache)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]:
```Scala
          new InMemoryFileIndex(
            sparkSession, globbedPaths, options, Some(partitionSchema), fileStatusCache)
```

This is also valid

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 3]
```diff
@@ -122,7 +122,7 @@ case class DataSource(
         val qualified = hdfsPath.makeQualified(fs.getUri, fs.getWorkingDirectory)
         SparkHadoopUtil.get.globPathIfNecessary(qualified)
       }.toArray
-      new InMemoryFileIndex(sparkSession, globbedPaths, options, None)
+      new InMemoryFileIndex(sparkSession, globbedPaths, options, None, fileStatusCache)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: This also impacts the streaming code path. If it is fine to streaming, the code changes look good to me.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 4]
```diff
@@ -86,7 +86,7 @@ case class DataSource(
   lazy val providingClass: Class[_] = DataSource.lookupDataSource(className)
   lazy val sourceInfo: SourceInfo = sourceSchema()
   private val caseInsensitiveOptions = CaseInsensitiveMap(options)
-
+  private lazy val fileStatusCache = FileStatusCache.getOrCreate(sparkSession)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: what's the life time of this cache?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 5]
```diff
@@ -364,7 +364,12 @@ case class DataSource(
             catalogTable.get,
             catalogTable.get.stats.map(_.sizeInBytes.toLong).getOrElse(defaultTableSize))
         } else {
-          new InMemoryFileIndex(sparkSession, globbedPaths, options, Some(partitionSchema))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: I'd like to create file status cache as a local variable, pass it to `getOrInferFileFormatSchema`, then use it here. It's much easier to reason about the lifetime of this cache by this way.

Comment (windpiger) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: ok, I think it is more reasonable~ thanks~

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala (hunk 3 above)
Comment (windpiger) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: I have make the it local only in the no streaming FileFormat match case~

**Comment by gatorsmile** [comment]
Why you closed it?

**Comment by windpiger** [comment]
oh...sorry ， I don't know when I close it...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 6]
```diff
@@ -109,7 +109,9 @@ case class DataSource(
    * @return A pair of the data schema (excluding partition columns) and the schema of the partition
    *         columns.
    */
-  private def getOrInferFileFormatSchema(format: FileFormat): (StructType, StructType) = {
+  private def getOrInferFileFormatSchema(
+      format: FileFormat,
+      fileStatusCache: FileStatusCache = NoopCache): (StructType, StructType) = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Please update the function description with a new `@parm`

**Comment by gatorsmile** [comment]
Please remove `[FOLLOW-UP]` from the PR title. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala (hunk 6 above)
Comment (windpiger) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: ok, thanks~

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 7]
```diff
@@ -106,6 +106,7 @@ case class DataSource(
    *     be any further inference in any triggers.
    *
    * @param format the file format object for this DataSource
+   * @param fileStatusCache fileStatusCache for InMemoryFileIndex
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: `@param fileStatusCache the shared cache for file statuses to speed up listing`

Comment (windpiger) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: ok thanks!

**Comment by gatorsmile** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**